### PR TITLE
Fixes #18770: Add API tests for rules

### PR DESCRIPTION
**Pull request by fanf**
https://issues.rudder.io/issues/18770

Add unit tests for Rule API. 

The test are based on a yaml file in `webapp/sources/rudder/rudder-rest/src/test/resources/api/api_rules.yml`. 
The format is quite simple with a request/response pair. 
It is very quick to add a new test case: just add more request/response pair in that file, or create a new `yml` file in `webapp/sources/rudder/rudder-rest/src/test/resources/api/`, it will automatically be run by `TestRestFromFileDef`.

The VERY big problem was to create mock repositories for rule, and directives, and techniques. A bit long, but now everything is in place to be able to do the same for nodes, parameters, ...
The mock services and test data are located:
- in `webapp/sources/rudder/rudder-rest/src/test/resources/configuration-repository` for the techniques (because since they need to be stored in a git repos, we need to have files at some point. They are copied during `/tmp/` during tests, and a git repos is created from there)
- in `webapp/sources/rudder/rudder-rest/src/test/scala/com/normation/rudder/MockServices.scala` for directives/rules and related repositories
- a lot of fake service init is done in `webapp/sources/rudder/rudder-rest/src/test/scala/com/normation/rudder/rest/RestTestSetUp.scala`

This PR also add a basic unserialisation test for Rule JSON, so that we can evolve it in the future (for example if we want to change json lib).

Finally, this PR move `copyReferenceLib` (the logic that init directive library) from boostrap to a standalone service, so that it can be used in mock repo too.

### CODE REVIEW

**Comment by fanf** [comment]
PR rebased

**Comment by fanf** [comment]
PR rebased

File: webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/RestExtractorService.scala [hunk 1]
```diff
@@ -313,7 +313,8 @@ final case class RestExtractorService (
     }
   }
 
-  private[this] def toRuleTarget(json:JValue, key:String ): Box[Option[RuleTarget]] = {
+  def toRuleTarget(json:JValue, key:String ): Box[Option[RuleTarget]] = {
+    println(s"**** extract '${key}' from ${json}")
```
Comment (VinceMacBuche) [inline webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/RestExtractorService.scala:317]: println here

File: webapp/sources/rudder/rudder-rest/src/test/scala/com/normation/rudder/MockServices.scala [hunk 2]
```diff
@@ -0,0 +1,1254 @@
...
+  val t3 = System.currentTimeMillis()
+  println(s"System Var Spec service : ${t3-t2bis} ms")
+
+  val techniqueParser: TechniqueParser = new TechniqueParser(
+      variableSpecParser
+    , new SectionSpecParser(variableSpecParser)
+    , systemVariableServiceSpec
+  )
+  val t4 = System.currentTimeMillis()
+  println(s"Technique parser        : ${t4-t3} ms")
```
Comment (VinceMacBuche) [inline webapp/sources/rudder/rudder-rest/src/test/scala/com/normation/rudder/MockServices.scala:145]: Do you want to keep println here ? 

I have no idea, it's in test it does no harm

**Comment by fanf** [comment]
PR rebased

**Review by VinceMacBuche — APPROVED** [review]

**Comment by fanf** [comment]
OK, merging this PR

**Pull request merged**